### PR TITLE
History: Add history routes in /me

### LINKED ISSUE #476: Implement /me/history route
Tracking issue for everything that relates to the `/me/history` route of the public API.  
See also [the API spec](https://github.com/codimd/server/blob/develop/docs/dev/public_api.yml) for details.

1. Add E2E tests:
- [ ] GET /me​/history
- [ ] GET ​/me​/history​/{note}
- [ ] PUT ​/me​/history​/{note}
- [ ] DELETE ​/me​/history​/{note}
2. Add implementation:
- [ ] GET /me​/history
- [ ] GET ​/me​/history​/{note}
- [ ] PUT ​/me​/history​/{note}
- [ ] DELETE ​/me​/history​/{note}

### PR DESCRIPTION
**Pull request by DerMolly**
### Component/Part
public API

### Description
This PR adds the whole complex of history to the public api, including
- history service and usage
- history entry and repo
- unit tests
- e2e test 

### Steps

- [x] Added implementation
- [x] Added / updated tests
- [x] I read the [contribution documentation](https://github.com/hedgedoc/hedgedoc/blob/develop/CONTRIBUTING.md) and
  made sure that:
  - My commits are signed-off to accept the DCO.
  - This PR targets the correct branch: `master` for 1.x & docs, `develop` for 2.x

### Related Issue(s)
fixes #476

### CODE REVIEW

File: src/auth/mock-auth.guard.ts [hunk 1]
```diff
@@ -16,7 +16,13 @@ export class MockAuthGuard {
   async canActivate(context: ExecutionContext) {
     const req = context.switchToHttp().getRequest();
     if (!this.user) {
-      this.user = await this.usersService.createUser('hardcoded', 'Testy');
+      // this assures that we can create the user 'hardcoded', if we need them before any calls are made or
```
Comment (davidmehren) [inline src/auth/mock-auth.guard.ts:19]: Please move this fix into a separate commit

File: src/history/history-entry.enity.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  PrimaryGeneratedColumn,
+  UpdateDateColumn,
+} from 'typeorm';
+import { User } from '../users/user.entity';
+import { Note } from '../notes/note.entity';
+
+@Entity()
+export class HistoryEntry {
+  @PrimaryGeneratedColumn('uuid')
+  id: string;
```
Comment (davidmehren) [inline src/history/history-entry.enity.ts:20]: Why do these need their own id? Shouldn't the composite key with `noteId` and `userId` suffice?

File: src/history/history-entry.enity.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  @ManyToOne((_) => Note, (note) => note.historyEntries, {
+    onDelete: 'CASCADE',
+  })
+  note: Note;
+
+  @Column()
+  pinStatus: boolean;
+
+  @UpdateDateColumn()
+  updatedAt: Date;
```
Comment (davidmehren) [inline src/history/history-entry.enity.ts:36]: This seems to be missing from the PlantUML schema

File: src/history/history-entry.enity.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+  public static create(user: User, note?: Note): HistoryEntry {
+    const newHistoryEntry = new HistoryEntry();
+    newHistoryEntry.user = user;
+    if (note) {
```
Comment (davidmehren) [inline src/history/history-entry.enity.ts:41]: A HistoryEntry without a note does not make much sense to me. Why do we make it optional?

(The column is also missing a nullable annotation, *if* `note` needs to be optional)

File: src/history/history.service.ts [hunk 5]
```diff
@@ -8,90 +8,101 @@ import { Injectable } from '@nestjs/common';
...
+
+  private async getEntry(
+    noteIdOrAlias: string,
+    user: User,
+  ): Promise<HistoryEntry> {
+    const note = await this.notesService.getNoteByIdOrAlias(noteIdOrAlias);
+    return await this.getEntryWithNote(note, user);
+  }
+
+  private async getEntryWithNote(
```
Comment (davidmehren) [inline src/history/history.service.ts:47]: We use `By` instead of `With` in other places, please rename.
To make the distinction between this function and `getEntry` more apparent, maybe rename `getEntry` to `getEntryByNoteIdOrAlias`?

File: src/history/history.service.ts [hunk 6]
```diff
@@ -8,90 +8,101 @@ import { Injectable } from '@nestjs/common';
...
-          updateUser: {
-            displayName: 'foo',
-            userName: 'fooUser',
-            email: 'foo@example.com',
-            photo: '',
-          },
-          viewCount: 42,
-        },
-        pinStatus: false,
+  async getUserHistory(user: User): Promise<HistoryEntry[]> {
```
Comment (davidmehren) [inline src/history/history.service.ts:32]: To improve consistency, might `getEntriesByUser` be better name?

File: test/public-api/me.e2e-spec.ts [hunk 7]
```diff
@@ -0,0 +1,158 @@
+/*
```
Comment (davidmehren) [inline test/public-api/me.e2e-spec.ts:1]: Please split renaming this file into a separate commit

File: src/history/history-entry.enity.ts [hunk 8]
```diff
@@ -0,0 +1,47 @@
+/*
```
Comment (davidmehren) [inline src/history/history-entry.enity.ts:1]: typo in filename

**Review by davidmehren — CHANGES_REQUESTED** [review]

File: src/api/public/notes/notes.controller.ts [hunk 9]
```diff
@@ -57,6 +59,22 @@ export class NotesController {
     );
   }
 
+  @UseGuards(TokenAuthGuard)
+  @Get(':noteIdOrAlias')
+  async getNote(@Request() req, @Param('noteIdOrAlias') noteIdOrAlias: string) {
+    // ToDo: check if user is allowed to view this note
+    try {
+      const note = await this.noteService.getNoteByIdOrAlias(noteIdOrAlias);
+      await this.historyService.createOrUpdateHistoryEntry(note, req.user);
```
[the remark this answers is not in the document]
Comment (DerMolly) [reply src/api/public/notes/notes.controller.ts:68]: No this is working as intended.

File: src/history/history-entry.enity.ts (hunk 2 above)
Comment (DerMolly) [reply src/history/history-entry.enity.ts:20]: Yeah this is probably not necessary.

File: src/history/history-entry.enity.ts (hunk 3 above)
Comment (DerMolly) [reply src/history/history-entry.enity.ts:36]: I'll add this.

File: src/history/history-entry.enity.ts (hunk 4 above)
Comment (DerMolly) [reply src/history/history-entry.enity.ts:41]: No this quite right. This is needed because of the way TypeORM works.
The create function without a note is used, when a note is created, in that case TypeORM adds the appropriate noteId to the database.

File: src/history/history-entry.enity.ts (hunk 8 above)
Comment (DerMolly) [reply src/history/history-entry.enity.ts:1]: done

File: src/history/history-entry.enity.ts (hunk 3 above)
Comment (DerMolly) [reply src/history/history-entry.enity.ts:36]: done

File: src/history/history-entry.enity.ts (hunk 2 above)
Comment (DerMolly) [reply src/history/history-entry.enity.ts:20]: done

File: src/history/history-entry.enity.ts (hunk 4 above)
Comment (DerMolly) [reply src/history/history-entry.enity.ts:41]: I added a comment

File: src/history/history.service.ts (hunk 5 above)
Comment (DerMolly) [reply src/history/history.service.ts:47]: done

File: src/history/history.service.ts (hunk 6 above)
Comment (DerMolly) [reply src/history/history.service.ts:32]: done

File: src/auth/mock-auth.guard.ts (hunk 1 above)
Comment (DerMolly) [reply src/auth/mock-auth.guard.ts:19]: done

File: test/public-api/me.e2e-spec.ts (hunk 7 above)
Comment (DerMolly) [reply test/public-api/me.e2e-spec.ts:1]: done

**Review by InnayTool — APPROVED** [review]

**Review by InnayTool — APPROVED** [review]

**Review by davidmehren — APPROVED** [review]
